### PR TITLE
Refactor TensorIterator to do allocations via MetaBase::set_output

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48687 Delete NativeFunctions.h include from Functions.h
* #48683 Move var and std overloads to Functions.cpp and remove native:: reference
* **#48659 Refactor TensorIterator to do allocations via MetaBase::set_output**
* #48195 Move argument grouping into FunctionSchema
* #48182 Refactor argument fields in FunctionSchema to Arguments
* #48116 Structured kernels generate Meta registrations

Detailed RFC at
https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md#handling-tensoriterator

What this diff does:
* Refactor allocation of outputs in TensorIterator into a call to a single function TensorIterator::set_output.  This nicely centralizes restriding logic and mostly eliminates the need for a separate named tensor propagation pass. The one exception is for inplace operations (`add_`), where previously we never actually call `set_output` when we determine resizing is not necessary; there's an extra propagate names in `allocate_or_resize_outputs` to handle this case (I audited all other `set_output` sites and found that we always hit this path in that situation). Although hypothetically this could cause problems for structured kernels (which require a `set_output` call in all cases), this codepath is irrelevant for structured kernels as a TensorIterator will never be constructed with an explicit out argument (remember, structured kernels handle out/functional/inplace variants). There's also a tricky case in `compute_types`; check the comments there for more details.
* Split TensorIterator into a TensorIteratorBase, which contains most of the logic but doesn't define `set_output`. A decent chunk of the diff is just the mechanical rename of TensorIterator to TensorIteratorBase. However, there are a few cases where we create fresh TensorIterator objects from another TensorIterator. In those cases, we always construct a fresh TensorIterator (rather than preserving the subclass of TensorIteratorBase that induced this construction). This makes sense, because a structured function class will contain metadata that isn't relevant for these downstream uses. This is done by *intentionally* permitting object slicing with the `TensorIterator(const TensorIteratorBase&)` constructor.
* Introduce a new `MetaBase` class which contains the canonical virtual method definition for `set_output`. This will allow structured classes to make use of it directly without going through TensorIterator (not in this PR).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25261844](https://our.internmc.facebook.com/intern/diff/D25261844)

### CODE REVIEW

File: aten/src/ATen/native/TensorIterator.cpp [hunk 1]
```diff
@@ -402,10 +402,25 @@ void TensorIterator::compute_types(const TensorIteratorConfig& config) {
     if (common_device == kCPU) {
       // Casts to outputs by creating temporaries of the correct dtype (if needed)
       if (config.cast_common_dtype_to_outputs_ && op.is_output && op.current_dtype != common_dtype_) {
+        TORCH_INTERNAL_ASSERT(op.tensor.defined());
```
Comment (bhosmer) [inline aten/src/ATen/native/TensorIterator.cpp:405]: I'm seeing the `continue` on line 372... not seeing how we get here with `op.tensor` undefined, but maybe I'm missing it

File: aten/src/ATen/native/TensorIterator.cpp [hunk 2]
```diff
@@ -1307,14 +1271,40 @@ void TensorIterator::build(TensorIteratorConfig& config) {
   view_offsets_ = DimVector(ndim_offsets, 0);
 }
 
-SplitUntil32Bit TensorIterator::with_32bit_indexing() const {
+void TensorIterator::set_output(int64_t output_idx, IntArrayRef sizes, IntArrayRef strides, TensorOptions options, DimnameList names) {
+  auto& op = operands_[output_idx];
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(output_idx < num_outputs_);
+  if (!op.tensor.defined()) {
+      TORCH_INTERNAL_ASSERT(op.is_type_defined(), "no type for operand", output_idx);
```
Comment (bhosmer) [inline aten/src/ATen/native/TensorIterator.cpp:1278]: On a quick read it looks like a less aggressive factoring would avoid some redundant tests when this is called from `allocate_or_resize_outputs`, at the cost of leaving more code in `fast_set_up` - but I'm assuming you figure the slow path can afford it (or that there's some API requirement that it be done in `set_output` that I'm missing)

File: aten/src/ATen/native/TensorIterator.h [hunk 3]
```diff
@@ -478,7 +488,9 @@ class CAFFE2_API TensorIteratorConfig final {
   // It would be better if this was && qualified, but this would be at the cost
   // of a lot of boilerplate above
   TensorIterator build() {
-    return TensorIterator(*this);
+    TensorIterator iter;
```
Comment (bhosmer) [inline aten/src/ATen/native/TensorIterator.h:491]: did you kill the `TensorIterator(TensorIteratorConfig&)` constructor just to clean things up?

**Review by ngimel — APPROVED** [review]
lgtm

**Review by bhosmer — APPROVED** [review]
Looks good! 

Only real question is about some minorly redundant logic that I think fell out of the refactoring straddle between `allocate_or_resize_outputs` and `fast_set_up` call sites. But I'd buy that this is the right sweet spot, at least for now.

File: aten/src/ATen/native/TensorIterator.h (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/TensorIterator.h:491]: You are right, I don't think I was supposed to delete this constructor, and I ended up just refactoring around the problem. Not sure if it's worth restoring it though!

Comment (bhosmer) [reply aten/src/ATen/native/TensorIterator.h:491]: Yeah I think this is better - more transparent. The other just hid the build call

File: aten/src/ATen/native/TensorIterator.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/native/TensorIterator.cpp:1278]: This is a good question. When I wrote this to start, I tried to assume as few preconditions on the call-site as possible, ergo the redundant tests. If I understand you correctly, it's just this assert that could be removed (I don't think any of the other asserts are removable). Seems reasonable.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/TensorIterator.cpp:405]: You can't get here with `op.tensor()` being undefined, that's why it's an assert :) Actually I missed the continue on line 372, but I was fairly confident (for other reasons) that it was impossible, but not 100% so.

**Comment by ezyang** [comment]
@bhosmer I merged this because I got a clean merge window but I still owe you updates for your comments. Coming soon.

**Pull request closed without merging**